### ISSUE TITLE
Add MPI build system

### ISSUE THREAD

**Issue by wds15** [stan-dev/math#778]
#### Summary:
Add MPI to makefile system

#### Description:
Along the lines of the GPU code, the MPI code will be build as:

- Users must include in `make/local` a STAN_MPI=true line (the actual defined value is ignored; it suffices to define the variable)
- In the libraries makefile the lib/boost/staged/lib/mpi.so is defined as a target file which is build by the boost build system. This target builds the libboost_mpi and its dependencies with it. Note that the mpi.so always has an "so" ending on Linux and Mac. The libboost_mpi has an ending .dylib or .so depending on the OS being MacOS or Linux, respectively
- Note that on MacOS additional steps are done in order to enable hard-coded dynamic linker paths.

#### Reproducible Steps:
NA

#### Current Output:
NA

#### Expected Output:
NA

#### Additional Information:
Provide any additional information here.

#### Current Version:
v2.17.0

**Comment by wds15** [comment]
if you want to try things out, just do a 

```
touch test/unit/fake_mpi_test.cpp
make test/unit/fake_mpi_test
```

which should trigger everything.

**Comment by yizhang-cae** [comment]
@bob-carpenter, @wds15, I suggest use `mpiexec` instead of `mpirun` for testing and running mpi applications. `mpirun` is shipped with many implementations but is not part of MPI-2 standard. The standard specifies `mpiexec` for better portability, see for example chapter 5 of [MPICH2 manual.](https://www.mpich.org/static/downloads/3.2.1/mpich-3.2.1-userguide.pdf) Therefore unless we specify the MPI implementations that Stan supports, I'd say we stay with the standards.

**Comment by bob-carpenter** [comment]
The doc @wds15 created makes it clear that we only test mpich.  To me, that means we only support mpich.   Given that we have no support contracts with anyone, it's all a bit moot because there's no real definition of "support".  We're not promising anyone anything.

**Comment by yizhang-cae** [comment]
As long as it's specified I'm ok. Good luck to those cluster admins.

**Comment by wds15** [comment]
fixed with #783
